fish-rust/src/wutil/format/printf.rs

```diff
@@ -0,0 +1,126 @@
+// Adapted from https://github.com/tjol/sprintf-rs
+// License follows:
+//
+// Copyright (c) 2021 Thomas Jollans
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is furnished
+// to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
+// OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF
+// OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+pub use super::format::Printf;
+use super::parser::{parse_format_string, ConversionType, FormatElement, NumericParam};
+use crate::wchar::{wstr, WString};
+
+/// Error type
+#[derive(Debug, Clone, Copy)]
+pub enum PrintfError {
+    /// Error parsing the format string
+    ParseError,
+    /// Incorrect type passed as an argument
+    WrongType,
+    /// Too many arguments passed
+    TooManyArgs,
+    /// Too few arguments passed
+    NotEnoughArgs,
+    /// Other error (should never happen)
+    Unknown,
+}
```

Comment:
there is a useful crate called `thiserror` that may help to improve error handling. The idea is that it will display a custom message alongside the error: https://docs.rs/thiserror/latest/thiserror/ 

```suggestion
use thiserror::Error;

/// Error type
[derive(Debug, Clone, Copy)]
pub enum PrintfError {
    #[error("Error parsing the format string")]
    ParseError,
    #[error("Incorrect type passed as an argument")]
    WrongType,
    #[error("Too many arguments passed")]
    TooManyArgs,
    #[error(" Too few arguments passed")]
    NotEnoughArgs,
    #[error("Other error (should never happen)")]
    Unknown,
}
```